src/R/Support/Impl/Help/Packages/PackageIndex.cs

```diff
@@ -2,106 +2,157 @@
+using System.Diagnostics;
 using System.Linq;
+using System.Threading.Tasks;
 using Microsoft.Common.Core;
 using Microsoft.Common.Core.Shell;
-using Microsoft.R.Support.Help.Definitions;
+using Microsoft.Common.Core.Threading;
+using Microsoft.R.Components.PackageManager.Model;
+using Microsoft.R.Host.Client;
+using Newtonsoft.Json.Linq;
+using static System.FormattableString;
 
 namespace Microsoft.R.Support.Help.Packages {
     /// <summary>
-    /// Index of packages available from the R engine. Collection 
-    /// of packages installed by the engine is furhished by a static
-    /// provider while list of the user-installed packages and 
-    /// list of per-project packages (when PackRat is used) 
-    /// are supplied by the providers exported via MEF. 
+    /// Index of packages available from the R engine.
     /// </summary>
     [Export(typeof(IPackageIndex))]
-    public class PackageIndex : IPackageIndex {
+    public sealed class PackageIndex : IPackageIndex {
         private readonly ICoreShell _shell;
-        private readonly Dictionary<string, IPackageInfo> _packages = new Dictionary<string, IPackageInfo>();
-        private readonly Lazy<IFunctionIndex> _functionIndexLazy;
+        private readonly IIntellisenseRHost _host;
+        private readonly IFunctionIndex _functionIndex;
+        private readonly ConcurrentDictionary<string, PackageInfo> _packages = new ConcurrentDictionary<string, PackageInfo>();
+        private readonly BinaryAsyncLock _buildIndexLock = new BinaryAsyncLock();
+        private bool _ready;
 
-        private IEnumerable<Lazy<IPackageCollection>> _collections;
-        private IPackageCollection _basePackages;
+        public static readonly IEnumerable<string> PreloadedPackages = new string[]
+            { "base", "stats", "utils", "graphics", "datasets", "methods" };
 
         [ImportingConstructor]
-        public PackageIndex(ICoreShell shell) {
+        public PackageIndex(ICoreShell shell, IIntellisenseRHost host, IFunctionIndex functionIndex) {
             _shell = shell;
-            _functionIndexLazy = Lazy.Create(() => _shell.ExportProvider.GetExportedValue<IFunctionIndex>());
+            _host = host;
+            _functionIndex = functionIndex;
         }
 
+        #region IPackageIndex
         /// <summary>
-        /// Collection or packages installed with the R engine.
+        /// Collection of all packages (base, user and project-specific)
         /// </summary>
-        public IEnumerable<IPackageInfo> BasePackages {
-            get {
-                if (_basePackages == null)
-                    _basePackages = new BasePackagesCollection(_shell.ExportProvider.GetExportedValue<IFunctionIndex>());
-
-                return _basePackages.Packages;
+        public IEnumerable<IPackageInfo> Packages => _packages.Values;
+
+        public async Task BuildIndexAsync() {
+            _ready = await _buildIndexLock.WaitAsync();
+            if (!_ready) {
```

Comment:
You don't need to use `_ready` field here - the whole purpose of the `BinaryAsyncLock` is to combine checking and lock in one primitive